### PR TITLE
Expanded span printing

### PR DESCRIPTION
**Pull request by DanielJCampbell**
r? @nrc

### CODE REVIEW

File: src/libsyntax/codemap.rs [hunk 1]
```diff
@@ -1274,4 +1340,45 @@ mod tests {
...
+        let sp = Span {lo: BytePos(12), hi: BytePos(23), expn_id: id};
+        let sstr = cm.span_to_expanded_string(sp);
+        let res_str = format!("{}\n{}\n  {}\n  {}",
+                                  cm.span_to_snippet(root).unwrap_or("".to_string()),
+                                  cm.span_to_string(root),
+                                  cm.span_to_snippet(sp).unwrap_or("".to_string()),
+                                  cm.span_to_string(sp));
+        assert_eq!(sstr, res_str);
+    }
+}
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: nit: reinstate the newline at the end of the file

File: src/libsyntax/codemap.rs [hunk 2]
```diff
@@ -897,6 +961,8 @@ impl CodeMap {
         }
     }
 
+
+
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: nit: extra whitespace

File: src/libsyntax/codemap.rs [hunk 3]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
...
+            cur_span = new_span;
+            let new_str = self.span_to_string(cur_span);
+            let new_snip = self.span_to_snippet(cur_span).unwrap_or(
+                "Snippet unavailable (span malformed or source missing)".to_string());
+            strings.push(new_str.into_boxed_str());
+            strings.push(new_snip.into_boxed_str());
+        }
+
+        //Reverse the vector (so output is in order: root -> source),
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: nit: start comments with a space, e.g.: `// Reverse`

File: src/libsyntax/codemap.rs [hunk 4]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
                         hi.col.to_usize() + 1)).to_string()
     }
 
+    pub fn span_to_expanded_string(&self, sp: Span) -> String {
+        let src_str = self.span_to_string(sp);
+        let src_snip = self.span_to_snippet(sp).unwrap_or(
+            "Snippet unavailable (span malformed or source missing)".to_string());
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: prefer `to_owned` to `to_string` - it's faster. nit: prefer breaking at the `.` than at the opening paren.

File: src/libsyntax/codemap.rs [hunk 5]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
...
+        if sp.expn_id == NO_EXPANSION || sp.expn_id == COMMAND_LINE_EXPN {
+            return format!("{}\n{}", src_snip, src_str);
+        }
+
+        let mut strings = vec![src_str.into_boxed_str(), src_snip.into_boxed_str()];
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Why use `Box<str>` here?

File: src/libsyntax/codemap.rs [hunk 6]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
                         hi.col.to_usize() + 1)).to_string()
     }
 
+    pub fn span_to_expanded_string(&self, sp: Span) -> String {
+        let src_str = self.span_to_string(sp);
+        let src_snip = self.span_to_snippet(sp).unwrap_or(
+            "Snippet unavailable (span malformed or source missing)".to_string());
+        if sp.expn_id == NO_EXPANSION || sp.expn_id == COMMAND_LINE_EXPN {
+            return format!("{}\n{}", src_snip, src_str);
+        }
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Do you need this and the `if let` on line 826?

File: src/libsyntax/codemap.rs [hunk 7]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
                         hi.col.to_usize() + 1)).to_string()
     }
 
+    pub fn span_to_expanded_string(&self, sp: Span) -> String {
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Could you add a doc comment to the function saying what it does?

File: src/libsyntax/codemap.rs [hunk 8]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
...
+        while cur_span.expn_id != NO_EXPANSION && cur_span.expn_id != COMMAND_LINE_EXPN {
+            let old_span = cur_span;
+            let new_span = self.with_expn_info(old_span.expn_id,
+                                              |ei| ei.map(|ei| ei.call_site));
+
+            //No info attached to src span, return
+            if let None = new_span {
+                return format!("{}\n{}", &*strings[1], &*strings[0]);
+            }
+            let new_span = new_span.unwrap();
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Could use `match` rather than `if let` ... `.unwrap`.

File: src/libsyntax/codemap.rs (hunk 5 above)
Comment (DanielJCampbell) [inline src/libsyntax/codemap.rs]: Needed to use Box<str> inside the while block so the strings I made in it could survive beyond that scope, boxed those two so they'd match the type of the vector.

File: src/liblibc [hunk 9]
```diff
@@ -1 +1 @@
-Subproject commit c27f4167a7706baba1ddbb0a9ab320ec18d097db
+Subproject commit 8531cc11e196b7ae9072ae7b384a930ff7b00dfb
```
Comment (huonw) [inline src/liblibc]: Is this change on purpose?

File: src/libsyntax/codemap.rs (hunk 4 above)
Comment (nrc) [inline src/libsyntax/codemap.rs]: I feel like you shouldn't need this code and the same code inside the loop - maybe you can refactor the loop a little

File: src/libsyntax/codemap.rs [hunk 10]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
...
+
+            //This compacts recursive macros down into a single entry,
+            //By only pushing sequential spans with different callees
+            let old_fmt = self.with_expn_info(
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Do you have an example of where this is a problem? I don't come across recursive macros very often

File: src/libsyntax/codemap.rs [hunk 11]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
...
+        let mut is_snip = true;
+        let mut output = "".to_string();
+        let mut first = true;
+        for s in strings {
+            if !first {
+                output.push_str("\n");
+            }
+            first = false;
+            output.push_str(&indent);
+            output.push_str(&*s);
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: probably only need `&s`

File: src/libsyntax/codemap.rs [hunk 12]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
...
+            if !is_snip {
+                indent.push_str("  ");
+            }
+            is_snip = !is_snip;
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: I would keep two vecs - one of coords, one of snippets. You ought to be able to do this all with a bunch of iterators:

```
strs.iter().zip(snips.iter()).rev().map(|(s, snip)| format!("{}  {}", s, snip)).collect().join("\n")
```
```

File: src/libsyntax/codemap.rs (hunk 10 above)
Comment (DanielJCampbell) [inline src/libsyntax/codemap.rs]: https://danielkeep.github.io/tlborm/book/pat-push-down-accumulation.html is the sort of scenario I was trying to handle, where a macro is designed to call itself a lot, in different ways (so it can't just check for the same span location when deciding which calls to merge).

File: src/libsyntax/codemap.rs [hunk 13]
```diff
@@ -1274,4 +1340,45 @@ mod tests {
...
+
+    #[test]
+    fn t10() {
+        //Test span_to_expanded_string works in base case (no expansion)
+        let cm = init_code_map();
+        let span = Span {lo: BytePos(0), hi: BytePos(12), expn_id: NO_EXPANSION};
+        let sstr = cm.span_to_expanded_string(span);
+
+        assert_eq!(sstr, format!("{}\n{}", cm.span_to_snippet(span).unwrap(),
+                                 cm.span_to_string(span)));
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: I would just use a string literal, rather than constructing the string with `format!`

File: src/libsyntax/codemap.rs [hunk 14]
```diff
@@ -1274,4 +1340,45 @@ mod tests {
...
+        let info = ExpnInfo { call_site: call_site, callee: callee};
+        let id = cm.record_expansion(info);
+
+        let sp = Span {lo: BytePos(12), hi: BytePos(23), expn_id: id};
+        let sstr = cm.span_to_expanded_string(sp);
+        let res_str = format!("{}\n{}\n  {}\n  {}",
+                                  cm.span_to_snippet(root).unwrap_or("".to_string()),
+                                  cm.span_to_string(root),
+                                  cm.span_to_snippet(sp).unwrap_or("".to_string()),
+                                  cm.span_to_string(sp));
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Again, I'd just use a string literal here. I'd also like a more complex test case.

File: src/liblibc (hunk 9 above)
Comment (nrc) [inline src/liblibc]: You probably want to `git submodule update`

File: src/libsyntax/codemap.rs (hunk 5 above)
Comment (nrc) [inline src/libsyntax/codemap.rs]: You probably want to use `String` rather than `Box<str>`, they are morally equivalent and the former is more standard.

File: src/libsyntax/codemap.rs (hunk 10 above)
Comment (nrc) [inline src/libsyntax/codemap.rs]: I would err on the side of simplicity here - it's probably unlikely users will bump into this sort of macro often enough to warrant squashing the trace.

Comment (DanielJCampbell) [inline src/libsyntax/codemap.rs]: It shouldn't squash the trace - it just only prints one entry for that macro. It will still print all spans before and after. I'll clarify the comment

**Comment by nrc** [comment]
This looks good. I would perhaps try a recursive approach rather than iterative - since we expect traces to be fairly short and acyclic, and there is no pressure to be ultra-fast.

File: src/libsyntax/codemap.rs [hunk 15]
```diff
@@ -807,6 +807,70 @@ impl CodeMap {
...
+        if sp.expn_id == NO_EXPANSION || sp.expn_id == COMMAND_LINE_EXPN {
+            return format!("{}\n{}", src_snip, src_str);
+        }
+
+        let mut strings = vec![src_str.into_boxed_str(), src_snip.into_boxed_str()];
+        let mut cur_span = sp;
+        while cur_span.expn_id != NO_EXPANSION && cur_span.expn_id != COMMAND_LINE_EXPN {
+            let old_span = cur_span;
+            let new_span = self.with_expn_info(old_span.expn_id,
+                                              |ei| ei.map(|ei| ei.call_site));
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: You expand iteratively down the `call_site` branch, but only one step down the `callee` branch (below), it seems to me you should be iterating to the end of both branches (which would be easier if you were being recursive). At that point you probably need to think about indentation or something to distinguish the structure.

**Comment by DanielJCampbell** [comment]
Refactored to recurse down callee and call-site spans

File: src/libsyntax/codemap.rs [hunk 16]
```diff
@@ -807,6 +807,83 @@ impl CodeMap {
...
+    /// Returns a formatted string showing the expansion chain of a span
+    ///
+    /// Spans are printed in the following format:
+    ///
+    /// filename:start_line:col: end_line:col
+    /// snippet
+    ///   CALLEE:
+    ///   Callee span, or 'None'
+    ///   CALLSITE:
+    ///   Callsite span, or 'None'
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Formatting of the output:
* I would not use caps for `callee`/`callsite` - no need to draw attention to the headings
* I would truncate the snippet to the smaller of 50 chars/one line + `...`
* rather than printing `None`, how about not printing anything at all (i.e., no heading or span)

File: src/libsyntax/codemap.rs [hunk 17]
```diff
@@ -807,6 +807,83 @@ impl CodeMap {
...
+    ///
+    /// Callees and callsites are printed recursively, expanding into their
+    /// own callee/callsite spans. Each layer of recursion has an increased indent
+    pub fn span_to_expanded_string(&self, sp:Span, dent: &str) -> String {
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: s/dent/indent

File: src/libsyntax/codemap.rs [hunk 18]
```diff
@@ -807,6 +807,83 @@ impl CodeMap {
...
+        let mut indent = dent.to_owned();
+        let mut output = "".to_owned();
+        let src_str = self.span_to_string(sp);
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: s/src_str/span_str

File: src/libsyntax/codemap.rs [hunk 19]
```diff
@@ -807,6 +807,83 @@ impl CodeMap {
...
+        let src_snip = self.span_to_snippet(sp)
+            .unwrap_or("Snippet unavailable (span malformed or source missing)".to_owned());
+        output.push_str(&format!("{}{}\n{}{}\n", indent, src_str, indent, src_snip));
+
+        if sp.expn_id != NO_EXPANSION && sp.expn_id != COMMAND_LINE_EXPN {
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: style nit: could do `if ... { return output }` to avoid rightward drift

File: src/libsyntax/codemap.rs [hunk 20]
```diff
@@ -807,6 +807,83 @@ impl CodeMap {
...
+            let mut callee = self.with_expn_info(sp.expn_id,
+                                             |ei| ei.map_or(None, |ei| ei.callee.span.clone()));
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: why use `map_or` instead of `map` here?

File: src/libsyntax/codemap.rs [hunk 21]
```diff
@@ -807,6 +807,83 @@ impl CodeMap {
...
+            let mut callsite = self.with_expn_info(sp.expn_id,
+                                               |ei| ei.map(|ei| ei.call_site.clone()));
+            indent.push_str("  ");
+            output.push_str(&indent);
+            output.push_str("CALLEE:\n");
+
+            while callee.is_some() && self.match_callees(&sp, &callee.unwrap()) {
+                callee = self.with_expn_info(callee.unwrap().expn_id,
+                                       |ei| ei.map_or(None, |ei| ei.callee.span.clone()));
+            }
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: When skipping the callees/callsites like this, it is probably best to indicate this in the trace dump with `...` or something

File: src/libsyntax/codemap.rs [hunk 22]
```diff
@@ -1274,4 +1351,100 @@ mod tests {
...
+        let sstr = cm.span_to_expanded_string(end, "");
+        let res_str =
+            "blork2.rs:2:1: 2:12\nsecond line\n  CALLEE:\n  None\n  \
+             CALLSITE:\n  blork2.rs:1:1: 1:12\n  first line.\n    CALLEE:\n    \
+             blork.rs:2:1: 2:12\n    second line\n      CALLEE:\n      \
+             blork.rs:1:1: 1:12\n      first line.\n      CALLSITE:\n      \
+             blork.rs:1:1: 1:12\n      first line.\n    CALLSITE:\n    \
+             blork.rs:2:1: 2:12\n    second line\n      CALLEE:\n      \
+             blork.rs:1:1: 1:12\n      first line.\n      CALLSITE:\n      \
+             blork.rs:1:1: 1:12\n      first line.\n";
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: I would use a raw string rather than putting the newlines in the string - it will be easier to read.

**Comment by nrc** [comment]
I like this approach much better! Thanks for making the changes. I think all these comments are fairly minor.

File: src/libsyntax/codemap.rs (hunk 20 above)
Comment (DanielJCampbell) [inline src/libsyntax/codemap.rs]: Because callee spans are an Option Span. So to save having an Option<Option<Span>>, I use map_or so I can have an Option<Span> and treat it in a way consistent with callsite.

Comment (nrc) [inline src/libsyntax/codemap.rs]: does. `ei.and_then(|ei|, ei.callee...)` work?

**Comment by DanielJCampbell** [comment]
Committed the suggested changes.

File: src/libsyntax/codemap.rs [hunk 23]
```diff
@@ -807,6 +807,92 @@ impl CodeMap {
...
+            return output;
+        }
+
+        let mut callee = self.with_expn_info(sp.expn_id,
+                                             |ei| ei.and_then(|ei| ei.callee.span.clone()));
+        let mut callsite = self.with_expn_info(sp.expn_id,
+                                               |ei| ei.map(|ei| ei.call_site.clone()));
+
+        indent.push_str("  ");
+        let mut flag = true;
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: could you give `flag` a more descriptive name please?

File: src/libsyntax/codemap.rs [hunk 24]
```diff
@@ -1274,4 +1360,118 @@ mod tests {
...
+    Callsite:
+    ...
+    blork.rs:2:1: 2:12
+    second line
+      Callee:
+      blork.rs:1:1: 1:12
+      first line.
+      Callsite:
+      blork.rs:1:1: 1:12
+      first line.
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Could you wrap he snippets in back ticks please - it'll make it easier to distinguish the code snippets from the rest

**Comment by nrc** [comment]
Last two comments! r+ with those addressed.
